// @weilliptic/weil-sdk/dist/utils/common.js
import pako from 'pako';
export const defaultTransport = 'multipart';
export const isNonSenatePodId = (podId) => podId.startsWith('POD_');
export const isSenatePodId = (podId) => !isNonSenatePodId(podId);
export const constructRestEndpoint = (baseUrl) => {
    return [baseUrl, '/rest'].join('');
};
const apiRouteMap = {
    SmartContractCreator: '/contracts/upload_smartcontract',
    SmartContractExecutor: '/contracts/execute_smartcontract',
};
export const constructTransactionEndpoint = (baseUrl, contractType, transport = defaultTransport) => [
    baseUrl,
    apiRouteMap[contractType],
    transport === 'post' ? '_mobile' : undefined,
]
    .filter(Boolean)
    .join('');
export const gzip = async (input) => {
    // Step 1: Encode the string into a Uint8Array
    const encoder = new TextEncoder();
    const uint8Array = encoder.encode(input);
    return pako.gzip(uint8Array);
};
export const formatPayloadForPrint = (originalPayload) => {
    const payload = JSON.parse(JSON.stringify(originalPayload));
    const STRING_CHAR_LIMIT = 512;
    function cleanupForPrint(obj) {
        for (let k in obj) {
            if (typeof obj[k] === 'object') {
                cleanupForPrint(obj[k]);
            }
            else if (typeof obj[k] === 'string' &&
                obj[k].length > STRING_CHAR_LIMIT) {
                obj[k] = [
                    obj[k].substring(0, STRING_CHAR_LIMIT / 2 - 2),
                    obj[k].substring(obj[k].length - (STRING_CHAR_LIMIT / 2 - 2)),
                ].join(' [ … ] ');
            }
        }
        return obj;
    }
    return JSON.stringify(cleanupForPrint(payload), null, 2);
};
