// @weilliptic/weil-sdk/dist/schema.js
import { z, ZodError } from 'zod';
import { WeilError } from './errors.js';
export const Schema = {
    u8: z.number(),
    u16: z.number(),
    u32: z.number(),
    // TODO switch to bigint
    u64: z.number(),
    uint: z.number(),
    i8: z.number(),
    i16: z.number(),
    i32: z.number(),
    i64: z.number(),
    // TODO switch to bigint
    f8: z.number(),
    f16: z.number(),
    f32: z.number(),
    f64: z.number(),
    int: z.number(),
    bool: z.boolean(),
    string: z.string(),
    void: z.null(),
    array: z.array,
    args: z.object,
    tuple: z.tuple,
    record: z.record,
    object: z.object,
    enum: z.enum,
    union: z.union,
    result: (t, e) => z.union([
        z.object({
            Ok: t,
        }),
        z.object({
            Err: e,
        }),
    ]),
    option: z.optional,
};
export function parseSchema(schema, value) {
    try {
        return schema.parse(value);
    }
    catch (error) {
        if (error instanceof ZodError) {
            const message = z.prettifyError(error);
            throw new WeilError(message, {
                type: 'Validation',
            });
        }
        throw new WeilError(`An unexpected error occurred: ${error}`);
    }
}
