// @weilliptic/weil-sdk/dist/walletCommon.js
import { isNonSenatePodId, isSenatePodId } from './utils/common.js';
import { selectPod } from './pods/selectPod.js';
export class WeilWalletCommon {
    _axiosInstance;
    async getAxiosInstance() {
        return this._axiosInstance;
    }
    podsCache;
    constructor() { }
    async listPods() {
        const response = await (await this.getAxiosInstance()).post('', {
            request_id: '987654321',
            category: 'ClusterManagement',
            request_type: 'GetPodInfo',
        });
        return response.data.result.podDetails;
    }
    async submitTransaction(_params) {
        throw new Error('Method not implemented');
    }
    pods = {
        list: async () => {
            if (!this.podsCache) {
                const pods = await this.listPods();
                const podSortFn = (a, b) => a.counter - b.counter;
                const nodeSortFn = (a, b) => {
                    if (a.host === b.host) {
                        return 0;
                    }
                    return a.host > b.host ? 1 : -1;
                };
                this.podsCache = pods.toSorted(podSortFn).map(pod => ({
                    ...pod,
                    activeNodes: pod.activeNodes.toSorted(nodeSortFn),
                }));
            }
            return [...this.podsCache];
        },
        listNonSenate: async () => {
            return (await this.pods.list()).filter(pod => isNonSenatePodId(pod.podId));
        },
        listSenate: async () => {
            return (await this.pods.list()).filter(pod => isSenatePodId(pod.podId));
        },
    };
    transactions = {
        submit: async (params) => {
            return this.submitTransaction(params);
        },
    };
    contracts = {
        deploy: async (body, widl, { name = null, pods: podsSpecifier = 'default', upgrade = false, logo, author, description, organization, config, context, outcall = false, auditLog = false, }) => {
            const pods = await this.pods.list();
            let targetPods = selectPod(pods.map(pod => pod.podId), podsSpecifier);
            // console.log(`Deploying to pods: ${targetPods.join(', ')}`)
            return await Promise.all(targetPods.map(podId => this.transactions.submit({
                toAddress: '',
                contract: {
                    type: 'SmartContractCreator',
                    body,
                    name,
                    pod: podId,
                    widl,
                    upgrade,
                    logo,
                    author,
                    description,
                    organization,
                    config,
                    context,
                    outcall,
                    auditLog,
                },
            })));
        },
        execute: (contractAddress, method, params = {}, options = {}) => {
            const toAddress = typeof params.to_addr === 'string' ? params.to_addr : '';
            return this.transactions.submit({
                toAddress,
                contract: {
                    address: contractAddress,
                    arguments: params || {},
                    method,
                    type: 'SmartContractExecutor',
                    isNonBlocking: options.isNonBlocking ?? false,
                    shouldHideArgs: options.shouldHideArgs ?? true,
                },
                stream: options?.stream,
                headers: options?.headers,
            });
        },
    };
}
