// @weilliptic/weil-sdk/dist/cli/token.js
export const Token = ((wallet, contractAddress) => ({
    name: async () => await wallet.contracts.execute(contractAddress, 'name', {}),
    symbol: async () => await wallet.contracts.execute(contractAddress, 'symbol'),
    decimals: async () => await wallet.contracts.execute(contractAddress, 'decimals'),
    details: async () => await wallet.contracts.execute(contractAddress, 'details'),
    total_supply: async () => await wallet.contracts.execute(contractAddress, 'total_supply'),
    balance_for: async (addr) => await wallet.contracts.execute(contractAddress, 'balance_for', {
        addr,
    }),
    allowance: async (owner, spender) => await wallet.contracts.execute(contractAddress, 'allowance', {
        owner,
        spender,
    }),
    transfer: async (to_addr, amount) => await wallet.contracts.execute(contractAddress, 'transfer', {
        to_addr,
        amount,
    }),
    transfer_from: async (from_addr, to_addr, amount) => await wallet.contracts.execute(contractAddress, 'transfer_from', {
        from_addr,
        to_addr,
        amount,
    }),
    approve: async (spender, value) => await wallet.contracts.execute(contractAddress, 'approve', {
        spender,
        value,
    }),
}));
